pages/k8s/charm-vsphere-integrator.md

```diff
@@ -0,0 +1,226 @@
+  nav: "kubernetes/docs/shared/_side-navigation.md"
+context:
+  title: "vSphere integrator charm"
+  description: Kubernetes-master Charm reference
+keywords: kubernetes-master, charm, config
+tags: [reference]
+sidebar: k8smain-sidebar
+permalink: charm-vsphere-integrator.html
+layout: [base, ubuntu-com]
+toc: False
+---
+
+
+
+
+This charm acts as a proxy to VMware vSphere and provides an [interface][] to
+provide a set of credentials for a somewhat limited project user to the
+applications that are related to this charm.
+
+## Usage
+
+When on a vSphere cloud, this charm can be deployed, granted trust via Juju to
+access vSphere, and then related to an application that supports the
+[interface][].
+
+For example, [Charmed Kubernetes][] has support for this, and can be
+deployed with the following bundle overlay:
+
+```yaml
+applications:
+  vsphere-integrator:
+    charm: cs:~containers/vsphere-integrator
+    num_units: 1
+relations:
+  - ['vsphere-integrator', 'kubernetes-master']
+  - ['vsphere-integrator', 'kubernetes-worker']
+```
+
+Using Juju 2.4 or later:
+
+```bash
+juju deploy cs:charmed-kubernetes --overlay ./k8s-vsphere-overlay.yaml
+juju trust vsphere-integrator
+```
+
+To deploy with earlier versions of Juju, you will need to provide the cloud
+credentials via the `credentials` charm config option:
+
+```bash
+cat <<EOJ > /path/to/cloud.json
+{
+  "vsphere_ip": "a.b.c.d",
+  "user": "joe",
+  "password": "passw0rd",
+  "datacenter": "dc0"
+}
+EOJ
+
+juju config vsphere-integrator credentials="$(base64 /path/to/cloud.json)"
+```
+
+## Configuration
+
+This charm supports multiple config options that can be used to describe they
+vSphere environment.
+
+The only required option is `datastore`, as it is not included in the Juju
+credential that this charm relies on. By default, this is set to *datastore1*.
+This can be changed with:
+
+```bash
+juju config vsphere-integrator datastore='mydatastore'
+```
+
+You may also configure a *folder* and *resource pool path* for this charm.
+Details about these options can be found in the [vmware documentation][]:
+
+```bash
+juju config vsphere-integrator folder='juju-kubernetes' respool_path='foo'
+```
+
+As mentioned in the **Usage** section, `credentials` may be set with a
+base64-encoded json file. When set, this data will take precedent over all
+other methods of specifying credentials for this charm.
+
+If `credentials` is empty, there are config options for each key that
+constitute a Juju credential. These can be set with:
+
+```bash
+juju config vsphere-integrator \
+  vsphere_ip='a.b.c.d' \
+  user='joe' \
+  password='passw0rd' \
+  datacenter='dc0'
+```
+
+>Note: If any of the credential config options are set, they must all be set.
+
+When all of the credential config options are empty, this charm will fall
+back to the credential data it received with `juju trust vsphere-integrator`.
+
+
+<!-- CONFIG STARTS -->
+<!--AUTOGENERATED CONFIG TEXT - DO NOT EDIT -->
```

Comment:
I think I understand why this config section comes before the examples whereas the other charms have the config last, but it was a bit jarring when flipping between them to see the config reference mixed into the middle.  I don't know that there's a reasonable way to get that arrangement for this particular charm, though, because of the additional required config.